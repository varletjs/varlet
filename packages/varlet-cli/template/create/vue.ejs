<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
