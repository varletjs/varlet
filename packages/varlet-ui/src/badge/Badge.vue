<template>
  <div :class="classes(n(), n('$--box'))">
    <transition :name="n('$-badge-fade')">
      <span
        v-bind="$attrs"
        v-show="!hidden"
        :class="classes(n('content'), ...contentClass)"
        :style="{ background: color }"
      >
        <var-icon v-if="icon && !dot" :name="icon" size="10px" />
        <span v-else>{{ values }}</span>
      </span>
    </transition>
    <slot />
  </div>
</template>

<script lang="ts">
import VarIcon from '../icon'
import { toNumber } from '@varlet/shared'
import { computed, defineComponent } from 'vue'
import { props } from './props'
import { createNamespace } from '../utils/components'
import type { ComputedRef } from 'vue'

const { n, classes } = createNamespace('badge')

export default defineComponent({
  name: 'VarBadge',
  components: { VarIcon },
  inheritAttrs: false,
  props,
  setup(props, { slots }) {
    const contentClass: ComputedRef<Array<string | null | undefined>> = computed(() => {
      const { type, position, dot, icon } = props

      const positionBasic = slots.default && `${n('position')} ${n(`--${position}`)}`
      const dotClass = dot ? n('dot') : null
      const positionClass = getPositionClass()
      const iconClass = icon ? n('icon') : null

      return [n(`--${type}`), positionBasic, dotClass, positionClass, iconClass]
    })

    const values: ComputedRef<string | number> = computed(() => {
      const { dot, value, maxValue } = props

      if (dot) return ''

      if (value !== undefined && maxValue !== undefined && toNumber(value) > maxValue) return `${maxValue}+`

      return value
    })

    const getPositionClass = (): string | undefined => {
      const { position, dot } = props

      if (dot && position.includes('right')) return n('dot--right')

      if (dot && position.includes('left')) return n('dot--left')
    }

    return {
      n,
      classes,
      values,
      contentClass,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../icon/icon';
@import './badge';
</style>
