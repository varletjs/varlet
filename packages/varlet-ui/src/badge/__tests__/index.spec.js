import Badge from '..'
import VarBadge from '../Badge'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'

test('test badge plugin', () => {
  const app = createApp({}).use(Badge)
  expect(app.component(Badge.name)).toBeTruthy()
})

describe('test badge component props', () => {
  test('test badge type', () => {
    ;['default', 'primary', 'info', 'success', 'warning', 'danger'].forEach((type) => {
      const wrapper = mount(VarBadge, {
        props: { type },
      })

      expect(wrapper.find('.var-badge--' + type).exists()).toBe(true)
      wrapper.unmount()
    })
  })

  test('test badge dot', async () => {
    const wrapper = mount(VarBadge, {
      props: {
        dot: true,
      },
    })

    expect(wrapper.find('.var-badge__dot').exists()).toBe(true)
    await wrapper.setProps({ dot: false })
    expect(wrapper.find('.var-badge__dot').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test badge value', () => {
    ;[1, '2'].forEach(async (value) => {
      const wrapper = mount(VarBadge, {
        props: {
          value,
          dot: true,
        },
      })

      expect(wrapper.find('.var-badge__content').html()).toContain('<span></span>')
      await wrapper.setProps({ dot: false })
      expect(wrapper.find('.var-badge__content').html()).toContain(`<span>${value}</span>`)
      wrapper.unmount()
    })
  })

  test('test badge max value', async () => {
    const wrapper = mount(VarBadge, {
      props: {
        maxValue: 1,
        value: 2,
      },
    })

    expect(wrapper.find('.var-badge__content').html()).toContain('<span>1+</span>')
    await wrapper.setProps({ value: 1 })
    expect(wrapper.find('.var-badge__content').html()).toContain(`<span>1</span>`)
    wrapper.unmount()
  })

  test('test badge position', () => {
    ;['right-top', 'right-bottom', 'left-top', 'left-bottom'].forEach(async (position) => {
      const template = `
        <var-badge>
          <div style="width: 100px; height: 30px"></div>
        </var-badge>
      `
      const wrapper = mount({
        components: {
          [VarBadge.name]: VarBadge,
        },
        template,
      })

      await wrapper.setProps({ position })
      expect(wrapper.find('.var-badge__content').classes()).toContain('var-badge--' + position)
      wrapper.unmount()
    })
  })

  test('test badge color', () => {
    const wrapper = mount(VarBadge, {
      props: { color: 'red' },
    })

    expect(wrapper.find('.var-badge__content').attributes('style')).toContain('background: red;')
    wrapper.unmount()
  })

  test('test badge icon', () => {
    const wrapper = mount(VarBadge, {
      props: { icon: 'notebook' },
    })

    expect(wrapper.find('.var-icon').classes()).toContain('var-icon-notebook')
    wrapper.unmount()
  })
})

test('test badge default slots', () => {
  const wrapper = mount(VarBadge, {
    slots: {
      default: () => '',
    },
  })

  expect(wrapper.find('.var-badge__content').element.textContent).toBe('0')
  wrapper.unmount()
})

test('test badge value slots', () => {
  const wrapper = mount(VarBadge, {
    slots: {
      value: () => 'value',
    },
  })

  expect(wrapper.find('.var-badge__content').element.textContent).toBe('value')
  wrapper.unmount()
})
