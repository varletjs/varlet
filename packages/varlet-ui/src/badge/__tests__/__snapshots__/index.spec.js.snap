// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test badge example 1`] = `
"<div class=\\"example\\">
  <div class=\\"app-type\\">主题色徽标</div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--default\\"><span>0</span></span></transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--primary\\"><span>0</span></span></transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--info\\"><span>0</span></span></transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--warning\\"><span>0</span></span></transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--success\\"><span>0</span></span></transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger\\"><span>0</span></span></transition-stub>
  </div>
  <div class=\\"app-type\\">圆点徽标</div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger var-badge__dot var-badge__dot--right\\"><span></span></span></transition-stub>
  </div>
  <div class=\\"app-type\\">自定义内容标签</div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger\\"><span>badge</span></span></transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger\\"><span>hot</span></span></transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger\\"><span>66</span></span></transition-stub>
  </div>
  <div class=\\"app-type\\">最大值</div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger\\"><span>88</span></span></transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger\\"><span>99+</span></span></transition-stub>
  </div>
  <div class=\\"app-type\\">不同定位徽标</div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger var-badge__position var-badge--right-top\\"><span>0</span></span></transition-stub>
    <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default\\" style=\\"color: rgb(0, 150, 136); border-color: #009688;\\"><span class=\\"var-chip__text-normal\\">右上</span>
      <!--v-if--></span>
    </transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger var-badge__position var-badge--right-bottom\\"><span>0</span></span></transition-stub>
    <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default\\" style=\\"color: rgb(0, 150, 136); border-color: #009688;\\"><span class=\\"var-chip__text-normal\\">右下</span>
      <!--v-if--></span>
    </transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger var-badge__position var-badge--left-top\\"><span>0</span></span></transition-stub>
    <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default\\" style=\\"color: rgb(0, 150, 136); border-color: #009688;\\"><span class=\\"var-chip__text-normal\\">左上</span>
      <!--v-if--></span>
    </transition-stub>
  </div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger var-badge__position var-badge--left-bottom\\"><span>0</span></span></transition-stub>
    <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default\\" style=\\"color: rgb(0, 150, 136); border-color: #009688;\\"><span class=\\"var-chip__text-normal\\">左下</span>
      <!--v-if--></span>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">是否显示徽标</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--2\\">
    <!--v-if-->
    <div class=\\"var-button__content\\">点击改变状态</div>
  </button>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--danger var-badge__position var-badge--right-top\\"><span>0</span></span></transition-stub>
    <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default\\" style=\\"color: rgb(0, 150, 136); border-color: #009688;\\"><span class=\\"var-chip__text-normal\\">徽标</span>
      <!--v-if--></span>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">自定义徽标颜色</div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--default var-badge__position var-badge--right-top\\" style=\\"background: rgb(98, 0, 234);\\"><span>0</span></span></transition-stub>
    <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default\\" style=\\"color: rgb(0, 150, 136); border-color: #009688;\\"><span class=\\"var-chip__text-normal\\">徽标</span>
      <!--v-if--></span>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">自定义徽标图标</div>
  <div class=\\"var-badge var--box\\">
    <transition-stub><span class=\\"var-badge__content var-badge--default var-badge__position var-badge--right-top var-badge__icon\\" style=\\"background: rgb(98, 0, 234);\\"><i class=\\"var-icon var-icon--set var-icon-notebook\\" style=\\"transition: transform 0ms; font-size: 10px;\\"></i></span></transition-stub>
    <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default\\" style=\\"color: rgb(0, 150, 136); border-color: #009688;\\"><span class=\\"var-chip__text-normal\\">徽标</span>
      <!--v-if--></span>
    </transition-stub>
  </div>
</div>"
`;

exports[`test badge position 1`] = `
"<div class=\\"var-badge var--box\\">
  <transition-stub><span class=\\"var-badge__content var-badge--default var-badge__position var-badge--right-top\\"><span>0</span></span></transition-stub>
  <div style=\\"width: 100px; height: 30px;\\"></div>
</div>"
`;

exports[`test badge position 2`] = `
"<div class=\\"var-badge var--box\\">
  <transition-stub><span class=\\"var-badge__content var-badge--default var-badge__position var-badge--right-bottom\\"><span>0</span></span></transition-stub>
  <div style=\\"width: 100px; height: 30px;\\"></div>
</div>"
`;

exports[`test badge position 3`] = `
"<div class=\\"var-badge var--box\\">
  <transition-stub><span class=\\"var-badge__content var-badge--default var-badge__position var-badge--left-top\\"><span>0</span></span></transition-stub>
  <div style=\\"width: 100px; height: 30px;\\"></div>
</div>"
`;

exports[`test badge position 4`] = `
"<div class=\\"var-badge var--box\\">
  <transition-stub><span class=\\"var-badge__content var-badge--default var-badge__position var-badge--left-bottom\\"><span>0</span></span></transition-stub>
  <div style=\\"width: 100px; height: 30px;\\"></div>
</div>"
`;

exports[`test badge style 1`] = `
"<div class=\\"var-badge var--box\\">
  <transition-stub><span class=\\"var-badge__content var-badge--default var-badge__icon\\" style=\\"background: rgb(98, 0, 234);\\"><i class=\\"var-icon var-icon--set var-icon-fire\\" style=\\"transition: transform 0ms; font-size: 10px;\\"></i></span></transition-stub>
</div>"
`;
